src/content/reference/react-dom/components/form.md

```diff
@@ -0,0 +1,445 @@
+      {pending ? "Submitting..." : "Submit"}
+    </button>
+  );
+}
+
+function Form({ action }) {
+  return (
+    <form action={action}>
+      <Button />
+    </form>
+  );
+}
+
+export default function App() {
+  return <Form action={submitForm} />;
+}
+```
+
+```js actions.js hidden
+export async function submitForm(query) {
+    await new Promise((res) => setTimeout(res, 1000));
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "experimental",
+    "react-dom": "experimental",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+</Sandpack>  
+
+<Pitfall>
+
+The `useFormStatus` Hook must be called from a Component that is located inside `<form>`. You cannot call `useFormStatus` from the same Component that returns `<form>`.
+
+```js
+function Form() {
+  const { pending } = useFormStatus() // 🚩 `pending` will never be `true`
+  return <form action={submit}> //...
+}
+```
+
+Instead call `useFormStatus` from inside a Component that is located inside `<form>`.
+
+```js
+function Button(){
+    const { pending } = useFormStatus() // ✅ `pending` will be derived from the parent form state
+    return <button disabled={pending}> //...
+}
+
+function Form() {
+  return <form> <Button /> //...
+}
+```
+
+</Pitfall>
+
+### Optimistically updating form data {/*optimistically-updating-form-data*/}
+The `useOptimistic` Hook provides a way to optimistically update the user interface before a background operation, like a network request, completes. In the context of forms, this technique helps to make apps feel more responsive. When a user submits a form, instead of waiting for the server's response to reflect the changes, the interface is immediately updated with the expected outcome. 
+
+For example, when a user types a message into the form and hits the "Send" button, the `useOptimistic` Hook allows the message to immediately appear in the list with a "Sending..." label, even before the message is actually sent to a server. This "optimistic" approach gives the impression of speed and responsiveness. The form then attempts to truly send the message in the background. Once the server confirms the message has been received, the "Sending..." label is removed.
+
+<Sandpack>
+
+
+```js App.js
+import { experimental_useOptimistic as useOptimistic, useState } from "react";
+import { deliverMessage } from "./actions.js";
+
+function Thread({ messages, sendMessage }) {
+  const [optimisticMessages, addOptimisticMessage] = useOptimistic(
+    messages,
+    (state, newMessage) => [
+      ...state,
+      {
+        message: newMessage,
+        sending: true
+      }
+    ]
+  );
+
+  return (
+    <div>
+      {optimisticMessages.map((m, i) => (
+        <div key={i}>
+          {m.message}
+          <small>{m.sending ? " (Sending...)" : ""}</small>
+        </div>
+      ))}
+      <form
+        action={sendMessage}
+        onSubmit={async (e) => {
```

Comment:
This function feels really clunky. Is this really the recommended way to do optimistic updates for forms?

Reply:
https://github.com/facebook/react/pull/27453 and https://github.com/facebook/react/pull/27454 fix it to be a noop and to warn when called out of a transition.

Reply:
Fixed, thanks @sophiebits your suggestion was very helpful!